### DISCUSSION TITLE
creating-a-source-generator-part-1-creating-an-incremental-source-generator/

### DISCUSSION THREAD

**Discussion by giscus** [andrewlock/blog-comments#24] [Announcements]
# creating-a-source-generator-part-1-creating-an-incremental-source-generator/

In this post I walk through how to create a practical .NET 6 incremental source generator: an enum extensions class with a fast ToString() implementation

https://andrewlock.net/creating-a-source-generator-part-1-creating-an-incremental-source-generator/

**Comment by CyrusNajmabadi** [comment]
Hi there. One of the incremental source generator api authors here. If possible, if recommend adding a note (or followup post) mentioning that your can eliminate several steps (and have them run much more efficiently) if you use this new api: https://github.com/dotnet/roslyn/blob/7d64f6edcc0ee5fe4399a7966e1a6b2bd3ef9005/src/Compilers/Core/Portable/SourceGeneration/Nodes/SyntaxValueProvider_ForAttributeWithMetadataName.cs#L78

Thanks very much!

**Reply by andrewlock** [reply]
Hi Cyrus, thanks, it's been on my radar to write a post about this for a while! I'll add that note shortly. 

Looking back at this now, of more concern is how many bad behaviour's it's doing - returning `Syntax` nodes, including the `Compilation` in the pipeline, collecting all the results into one 😅 I'll definitely fix these too

**Comment by Miljankg** [comment]
Hi Andrew,

Thanks a lot for this tutorial. It helped me a lot. I do have one question regarding the pipeline you built. In the part where Syntax Provider is created (`context.SyntaxProvider.CreateSyntaxProvider`) there are two parameters passed, predicate and transform. You stated that the predicate should be efficient, and opted for brief and quick filtering by the number of attributes of a particular syntax node. Then in the transform, nodes are filtered further by looking for the particular attribute.

How much of a problem would it be if I would move the complete filtering in your example, into the predicate and leave the transform to do only that, getting and transforming the data to what is needed further down the pipeline? Would you consider such a predicate inefficient? Talking strictly about your example.

Thanks in advance!
Miljan

**Reply by Miljankg** [reply]
Just to provide some info on this in case it helps someone. The predicate serves for examining _Syntax_ Nodes in isolation. The name of the attribute is available here (e.g. SomeAttribute), but the namespace cannot be established from Syntax Node alone. Therefore, the predicate has only a simple filtering of Syntax Nodes.

For full attribute name (with namespace), Semantic Model is needed, where Symbols associated with Syntax Node can be examined. This is available in the transform method.

For more information on these visit:

- [Incremental Generators in GitHub](https://github.com/dotnet/roslyn/blob/main/docs/features/incremental-generators.md#syntaxvalueprovider)
- [This blogpost I found useful](https://joshvarty.com/2014/10/30/learn-roslyn-now-part-7-introducing-the-semantic-model/)
- [Andew's explanation of generator pipeline](https://andrewlock.net/creating-a-source-generator-part-1-creating-an-incremental-source-generator/#5-building-the-incremental-generator-pipeline)

**Comment by goldytech** [comment]
I’m unable to get this working on latest Rider version on MacOS. For me the source generator was able to generate code after I decorated my class with the marker attribute, even intellisense also worked. But when I try to run or compile the console app which consumes the source generator via project reference, it gives me error of “Cannot resolve the  method which is auto generated

**Comment by drdamour** [comment]
in section 3, what's the reason to define the enum as a source generated type instead of just a type in the assembly that can be referenced in downstream project?  i feel like i'm losing simple compiler/type checking stuff i'm familiar with by defining the attribute in a string, but i'm guessing there's an advantage to doing it that way.

**Reply by drdamour** [reply]
derp u answered this in https://andrewlock.net/creating-a-source-generator-part-7-solving-the-source-generator-marker-attribute-problem-part1/ & https://andrewlock.net/creating-a-source-generator-part-8-solving-the-source-generator-marker-attribute-problem-part2/

**Comment by andyolivares** [comment]
Hi Andrew:

First of all, thank you for these articles. They helped me to design my own source generator for a couple of parts of my system that heavily relied on Reflection. I've seen improvements of up to ~90x in speed at some places.

The "simple" generator (a single generator lib project) worked just fine, but since I splitted my generator into a few libraries to share code between 3 different generators, things went south and I've been unable to get it working again.

My project configuration is almost identical to yours, but VS2022 fails to generate any code nor compile (my generated classes are not being found).

I've been reading online and tried a few things. Most notably, I've been through some of the things mentioned here:
https://turnerj.com/blog/the-pain-points-of-csharp-source-generators

Added what they mention in Pain Point 3 and now my generator seems to create my classes (at least VS2022 compiles my project) but can't see any generated code in the IDE (as with the "simple" generator).

I'm not using the generator as a Nuget package but as a project reference for now. As I said before, the "simple" case being just one project, worked just fine, but when I separated the generator into different projects and started adding references between them, it stopped working.

Did you experience any of this? Do you know where to look at?

Thank you in advance!

Andy

**Reply by andrewlock** [reply]
Hey, short answer, I think you're going to be fighting with this approach - it's not something I've tried and honestly, I'm not going to try as I won't be surprised if it just doesn't work 😬 

If you don't need type unification, one thing you could try doing is using a shared project ([similar to this](https://learn.microsoft.com/en-us/xamarin/cross-platform/app-fundamentals/shared-projects?tabs=macos)), or adding the files as links. It doesn't really _solve_ your issue, but you may be able skirt around it!

**Reply by andyolivares** [reply]
Hi Andrew, thank you for your reply. At the end, I've managed to solve the issue and now my generators are doing their job pretty well. For anyone's reference I had to do something quite "hacky" (really don't know if that's the right approach, but it works):

```
<PropertyGroup>
  <TargetFramework>netstandard2.0</TargetFramework>
  <LangVersion>9.0</LangVersion>
  <IncludeBuildOutput>false</IncludeBuildOutput>
  <IsRoslynComponent>true</IsRoslynComponent>
</PropertyGroup>

<ItemGroup>
  <PackageReference Include="Microsoft.CodeAnalysis.Analyzers" Version="3.3.3" PrivateAssets="all" />
  <PackageReference Include="Microsoft.CodeAnalysis.CSharp" Version="4.3.0" PrivateAssets="all" />
  <ProjectReference Include="..\Generators.Common\Generators.Common.csproj" PrivateAssets="all" ExcludeAssets="runtime" />
</ItemGroup>

<PropertyGroup>
  <GetTargetPathDependsOn>$(GetTargetPathDependsOn);GetDependencyTargetPaths</GetTargetPathDependsOn>
</PropertyGroup>

<Target Name="GetDependencyTargetPaths">
  <ItemGroup>
    <TargetPathWithTargetPlatformMoniker Include="$(SolutionDir)src\Generators.Common\bin\$(Configuration)\netstandard2.0\Generators.Common.dll" IncludeRuntimeDependency="false" />
  </ItemGroup>
</Target>
```

As you can see, I had to include my shared project's DLL with **TargetPathWithTargetPlatformMoniker**. Not sure what it really does under the hood, but it did the trick.

Now, I'm having troubles trying to generate any code from **dotnet build** (as in VS2022 it works just fine).

Regards,

Andy

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at June 28 2022, 09:56, in reply to [Rich Armstrong](https://disqus.com/by/disqus_Kb8gaVoDN1/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/24#discussioncomment-6062892)</em>

---

<p>oops yeah, removed it now, thanks</p>

[a comment by an automated account is not shown]

**Comment by luizfbicalho** [comment]
Nice post , I always read it again to improve my generators

I will change my generators to use global namespaces and I got a point here

what is the best practice to generics and nullable?
```cs
global::System.Threading.Tasks.Task<global::System.String?> x = null;
global::System.Threading.Tasks.Task<string?> x2 = null;
global::System.Threading.Tasks.Task<global::System.Nullable<global::System.String>> x3 = null;
```

**Reply by CyrusNajmabadi** [reply]
use `string?`

**Reply by luizfbicalho** [reply]
Is there any extension or nuget package that format any type as global to use in source generators @andrewlock?

**Comment by RichardCrawshaw** [comment]
This for me is one of the most useful tutorials available. Thank you for writing it.

I'm now on my second Source Generator. The first I was able to get working fairly easily: it is driven by data from a CSV file and is all self contained.

The second is proving more problematical. This is driven by data in a JSON schema, which requires either Newtonsoft.Json or System.Text.Json to ingest the schema data; I've opted for the latter. What I'm finding is that there are a whole series of hoops that need to be jumped through to include the required NuGet package; specifically adding 
```
    <PackageReference Include="System.Text.Json" 
                      Version="9.0.2" PrivateAssets="all" GeneratePathProperty="true" />
```
and 
```
      <TargetPathWithTargetPlatformMoniker Include="$(PkgSystem_Text_Json)\lib\netstandard2.0\System.Text.Json.dll"
                                           IncludeRuntimeDependency="false" />
```
in the project file for the Source Generator.

Adding that pair isn't a particular pain; but what is is manually adding the same for all the packages that System.Text.Json references. It feels wrong to have to explicitly include the dependent packages, particularly when the VS NuGet Package Manager has now split implicitly added packages from those that have been explicitly added.

So, I'm wondering if there's something that I'm missing here.

**Reply by andrewlock** [reply]
The short answer is 
1. Luckily for me, none of my generates have required this 😅
2. Yes, you're pretty much doing it correctly (as [described in the cookbook](https://github.com/dotnet/roslyn/blob/main/docs/features/source-generators.cookbook.md#use-functionality-from-nuget-packages)), though that uses `<None>` with `Pack="true"` which I would recommend instead
3. Yes, it's a real pain, as you also need to pack all the transitive dependencies

**Reply by RichardCrawshaw** [reply]
Thank you for the confirmation. I'll look more deeply at `<None>` and `Pack="true"`.

Overall, this feels like a deficiency in Visual Studio, and I'm guessing that nobody has yet written an extension to plug the hole.

**Comment by voroninp** [comment]
Surprisingly, VS Solution explorer shows the folders `bin/Debug` when dll is added.
```
<None Include="$(OutputPath)\$(AssemblyName).dll" Pack="true" 
        PackagePath="analyzers/dotnet/cs" Visible="false" />
```
Yet the file is indeed hidden because of `Visible="false"`.

**Comment by NickStrupat** [comment]
Hi Andrew, love your blog posts. Just wondering:

Is the return type of `GetSemanticTargetForGeneration` correct? I think it should be `EnumToGenerate?` but I'm not sure.

**Reply by andrewlock** [reply]
Oops, yes, you're right, thanks!
